src/ring/middleware/oauth2.clj

```diff
@@ -188,11 +191,102 @@
+  (let [expired-access-token? (fn [[_ {:keys [expires refresh-token]}]]
+                                (and refresh-token expires
+                                     (.before expires (Date.))))]
+    (->> access-tokens (filter expired-access-token?) (into {}))))
+
+(defn- update-tokens [access-tokens [profile-key maybe-grant]]
+  (if maybe-grant
+    ;; `update ... merge` to properly handle case where authorization server
+    ;; does not update the refresh token after use and we should re-use the
+    ;; existing refresh token
+    (update access-tokens profile-key merge maybe-grant)
+    (dissoc access-tokens profile-key)))
+
+(def refresh-socket-timeout 60000)
+
+(defn- refresh-token-request-options [profile refresh-token]
+  (-> (token-http-options profile {:grant_type "refresh_token"
+                                   :refresh_token refresh-token})
+      (assoc :socket-timeout refresh-socket-timeout)))
+
+(defn- refresh-one-token
+  ([profile refresh-token]
+   (-> (http/request (refresh-token-request-options profile refresh-token))
+       format-access-token))
+  ([profile refresh-token respond raise]
+   (let [opts (-> (refresh-token-request-options profile refresh-token)
+                  (assoc :async? true))]
+     (http/request opts (comp respond format-access-token) raise))))
+
+(defn- refresh-tasks [profiles access-tokens]
+  (->> (expired-access-tokens access-tokens)
+       (keep (fn [[profile-key {:keys [refresh-token]}]]
+               (when (and (get profiles profile-key) refresh-token)
+                 [profile-key [(get profiles profile-key) refresh-token]])))))
+
+(defn- async-map-values [f respond m]
+  (let [total (count m)
+        results (atom {})
+        respond-when-done #(when (= (count %) total) (respond %))]
+    (if (zero? total)
+      (respond {})
+      (doseq [[k v] m
+              :let [respond #(respond-when-done (swap! results assoc k %))
+                    raise (fn [_] (respond nil))]]
+        (f v respond raise)))))
+
+(defn- refresh-all-tokens
+  ([profiles access-tokens]
+   (->> (refresh-tasks profiles access-tokens)
+        (map (fn [[profile-key [profile refresh-token]]]
+               [profile-key
+                (try (refresh-one-token profile refresh-token)
+                     (catch clojure.lang.ExceptionInfo _ nil))]))
+        (reduce update-tokens access-tokens)))
+  ([profiles access-tokens respond]
+   (async-map-values
+    (fn [[profile refresh-token] respond raise]
+      (refresh-one-token profile refresh-token respond raise))
+    (fn [refreshed-tokens]
+      (respond
+       (reduce update-tokens access-tokens refreshed-tokens)))
+    (refresh-tasks profiles access-tokens))))
+
+(defn- assoc-access-tokens-in-request [request tokens]
+  (if tokens
     (assoc request :oauth2/access-tokens tokens)
     request))
 
+(defn- assoc-access-tokens-in-response
+  [original-tokens updated-tokens response]
+  (if (and (not (contains? response :session))
```

Comment:
This doesn't look quite correct. I believe it should be:

```clojure
(and (not (and (contains? response :session)
               (nil? (:session response))))
     (not= original-tokens updated-tokens))
```

Or more clearly:

```clojure
(and (not (nil-session? response))
     (not= original-tokens updated-tokens))
```

In that the only time we don't want to update the response is if the tokens haven't changed, or the session has been set explicitly to `nil` and is therefore being deleted.

Reply:
Should be addressed in [26f2b34](https://github.com/weavejester/ring-oauth2/pull/56/commits/26f2b34d1176b5ef33b6b2814f4cbaa90bf1e5bf), with de morgan's law and branch order flipped